nose2/plugins/prettyassert.py

```diff
@@ -0,0 +1,360 @@
+        self.seen_tokens = collections.OrderedDict()
+
+        # the previous token seen as a tuple of (tok_type, token_name)
+        # (or None when we start)
+        self.last_tok = None
+
+    def get_token_collection(self):
+        return self.seen_tokens
+
+    def process(self, toktype, tok, start, end, line):
+        """
+        A tokenization processor for tokenize.generate_tokens
+        Skips certain token types, class names, etc
+
+        When an identifiable/usable token is found, add it to
+        collect_tokens
+        (which we add to extraDetail later)
+
+        When an "assert" statement is found, reset the token collection
+        and return the start line (relative to the text being tokenized)
+        """
+        prior_tok = self.last_tok
+        self.last_tok = (toktype, tok)
+
+        # CASE 0: skip non "NAME" or "OP" tokens and clear current resolution
+        #
+        # NAME is most identifiers and keywords
+        # OP is operators, including .
+        #
+        # special note: don't clear resolution for whitespace (e.g. newline)
+        if toktype not in (tokenize.NAME, tokenize.OP):
+            # only newline for now, maybe we'll find others
+            if toktype not in (tokenize.NEWLINE,):
+                self.doing_resolution = None
+            return
+
+        """
+        CASE 1: Operator token
+
+        skip tokens and either leave resolution in progress or reset, depending
+
+        continue resolution for
+          "."
+            because that's what attribute resolution *is*
+          ")"
+            this is handy, as it means that "(x).y" works
+
+        reset resolution for everything else, e.g. "[", "]", ":"
+        special note: reset resolution for "("
+
+        failing to filter out ")" can result in badness in cases like this:
+            >>> def foo():
+            >>>     return [1]
+            >>> foo.pop = 2
+            >>> ...
+            >>> def test_foo():
+            >>>    assert foo().pop() == 2
+
+        if we stop resolution when we see an LPAREN, we resolve `foo`
+        successfully, fail on `pop` and everything is OK, but if we try to
+        traverse the LPAREN, we get `foo.pop = 2` in our values, which is
+        wrong
+        """
+        if toktype == tokenize.OP:
+            if tok not in (".", ")"):
+                self.doing_resolution = None
+            return
+
+        # CASE 2: "assert" statement
+        # assert statement was reached, reset
+        # return the start line (start = (startrow, startcol))
+        if tok == 'assert':
+            self.seen_tokens.clear()
+            self.doing_resolution = None
+            return start[0]
+
+        # handle tokens
+
+        # CASE 3: a name is being resolved,
+        #         there is a previous token,
+        #         and it's a "." operator
+        if self.doing_resolution and prior_tok and (
+                prior_tok[0] == tokenize.OP and prior_tok[1] == '.'):
```

Comment:
maybe use a namedtuple for prior_tok? so reader doesn't have to scroll back an remember which is at index 0 vs 1.